File: android-app/feature/feature-player/src/main/java/tv/bayit/plus/feature/player/dialogue/SharedInteractionViewModel.kt
package tv.bayit.plus.feature.player.dialogue

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import tv.bayit.plus.core.common.logging.BayitLogger
import tv.bayit.plus.core.network.api.BayitApiClient
import tv.bayit.plus.core.network.websocket.ChannelType
import tv.bayit.plus.core.network.websocket.WebSocketConnection
import tv.bayit.plus.core.network.websocket.WebSocketManager
import javax.inject.Inject

/** ViewModel for shared interactive sessions within watch parties (WS4). */
@HiltViewModel
class SharedInteractionViewModel @Inject constructor(
    private val sharedApi: SharedInteractionApi,
    private val apiClient: BayitApiClient,
    private val webSocketManager: WebSocketManager,
    private val json: Json,
    private val logger: BayitLogger,
) : ViewModel() {
    private val _sessionId = MutableStateFlow<String?>(null)
    val sessionId: StateFlow<String?> = _sessionId.asStateFlow()
    private val _partyId = MutableStateFlow<String?>(null)
    val partyId: StateFlow<String?> = _partyId.asStateFlow()
    private val _isActive = MutableStateFlow(false)
    val isActive: StateFlow<Boolean> = _isActive.asStateFlow()
    private val _participants = MutableStateFlow<List<SharedParticipant>>(emptyList())
    val participants: StateFlow<List<SharedParticipant>> = _participants.asStateFlow()
    private val _currentTurnUserId = MutableStateFlow<String?>(null)
    val currentTurnUserId: StateFlow<String?> = _currentTurnUserId.asStateFlow()
    private val _turnCountdown = MutableStateFlow(0)
    val turnCountdown: StateFlow<Int> = _turnCountdown.asStateFlow()
    private val _exchanges = MutableStateFlow<List<DialogueExchangeItem>>(emptyList())
    val exchanges: StateFlow<List<DialogueExchangeItem>> = _exchanges.asStateFlow()
    private val _isSending = MutableStateFlow(false)
    val isSending: StateFlow<Boolean> = _isSending.asStateFlow()
    private val _characterName = MutableStateFlow<String?>(null)
    val characterName: StateFlow<String?> = _characterName.asStateFlow()
    private var countdownJob: Job? = null
    private var wsConnection: WebSocketConnection? = null

    fun startSharedSession(
        partyId: String,
        contentId: String,
        momentTimestamp: Double,
        characterName: String,
        profileId: String,
        avatarId: String,
        displayName: String,
    ) {
        viewModelScope.launch {
            try {
                val response = apiClient.safeApiCall {
                    sharedApi.startSharedInteraction(
                        partyId = partyId,
                        request = SharedStartRequest(
                            contentId = contentId,
                            momentTimestamp = momentTimestamp,
                            characterName = characterName,
                            profileId = profileId,
                            avatarId = avatarId,
                            displayName = displayName,
                        ),
                    )
                }
                _sessionId.value = response.sessionId
                _partyId.value = partyId
                _characterName.value = characterName
                _isActive.value = true
                _exchanges.value = emptyList()
                connectWebSocket(partyId, response.sessionId)
                logger.info(
                    "Shared session started",
                    mapOf("sessionId" to response.sessionId, "partyId" to partyId),
                )
            } catch (e: Exception) {
                logger.error("Failed to start shared session", error = e)
                _isActive.value = false
            }
        }
    }

    fun sendMessage(text: String, addressedCharacter: String? = null) {
        val sid = _sessionId.value ?: return
        val pid = _partyId.value ?: return
        if (text.isBlank()) return
        _isSending.value = true
        viewModelScope.launch {
            try {
                val response = apiClient.safeApiCall {
                    sharedApi.sendSharedMessage(
                        partyId = pid, sessionId = sid,
                        request = SharedMessageRequest(message = text, addressedCharacter = addressedCharacter),
                    )
                }
                val items = response.exchanges.map { ex ->
                    DialogueExchangeItem(
                        speaker = ex.speaker, messageText = ex.messageText,
                        characterName = ex.characterName, audioUrl = ex.audioUrl,
                        animatedVideoUrl = ex.animatedVideoUrl, participantName = ex.participantName,
                    )
                }
                _exchanges.value = _exchanges.value + items
                logger.debug("Shared message sent", mapOf("sessionId" to sid, "count" to items.size.toString()))
            } catch (e: Exception) {
                logger.error("Failed to send shared message", error = e, metadata = mapOf("sessionId" to sid))
            } finally {
                _isSending.value = false
            }
        }
    }

    fun endSession() {
        val sid = _sessionId.value ?: return
        val pid = _partyId.value ?: return
        viewModelScope.launch {
            try {
                apiClient.safeApiCall { sharedApi.endSharedInteraction(pid, sid) }
                logger.info("Shared session ended", mapOf("sessionId" to sid))
            } catch (e: Exception) {
                logger.error("Failed to end shared session", error = e)
            } finally { cleanup() }
        }
    }

    private fun connectWebSocket(partyId: String, sessionId: String) {
        viewModelScope.launch {
            try {
                val wsUrl = apiClient.retrofit.baseUrl().toString()
                    .replace("https://", "wss://").replace("http://", "ws://") +
                    "api/v1/ws/shared-interaction/$partyId/$sessionId"
                val conn = webSocketManager.connect(wsUrl, ChannelType.SHARED_INTERACTION)
                wsConnection = conn
                conn.messages.onEach { handleWsMessage(it) }.launchIn(viewModelScope)
            } catch (e: Exception) {
                logger.error("Failed to connect shared WS", error = e)
            }
        }
    }

    private fun handleWsMessage(text: String) {
        val parsed = try { json.decodeFromString<JsonObject>(text) } catch (e: Exception) {
            logger.error("Failed to parse shared WS message", error = e); return
        }
        when (parsed["type"]?.jsonPrimitive?.content) {
            "turn_change" -> {
                _currentTurnUserId.value = parsed["user_id"]?.jsonPrimitive?.content
                val duration = parsed["duration"]?.jsonPrimitive?.content?.toIntOrNull() ?: TURN_DURATION_SECONDS
                startCountdown(duration)
            }
            "session_ended" -> cleanup()
        }
    }

    private fun startCountdown(seconds: Int) {
        countdownJob?.cancel()
        _turnCountdown.value = seconds
        countdownJob = viewModelScope.launch {
            var remaining = seconds
            while (remaining > 0) { delay(COUNTDOWN_INTERVAL_MS); remaining--; _turnCountdown.value = remaining }
        }
    }

    private fun cleanup() {
        countdownJob?.cancel()
        wsConnection?.let { webSocketManager.disconnect(it.id) }
        wsConnection = null
        _sessionId.value = null
        _partyId.value = null
        _isActive.value = false
        _currentTurnUserId.value = null
        _turnCountdown.value = 0
        _exchanges.value = emptyList()
        _characterName.value = null
    }

    override fun onCleared() { cleanup(); super.onCleared() }
}

private const val TURN_DURATION_SECONDS = 30
private const val COUNTDOWN_INTERVAL_MS = 1000L

File: ios-app/BayitPlusApp/ViewModels/SharedInteractionViewModel.swift
import BayitCore
import Foundation
import Observation

/// Manages shared interactive sessions within watch parties.
/// Handles turn-based interaction, participant tracking, and event processing.
@MainActor
@Observable
final class SharedInteractionViewModel {
    // MARK: - Public State

    private(set) var sessionId: String?
    private(set) var participants: [SharedParticipant] = []
    private(set) var currentTurnUserId: String?
    private(set) var exchanges: [DialogueExchange] = []
    private(set) var turnsCompleted = 0
    private(set) var maxTurnsPerParticipant = 3
    private(set) var isSending = false
    private(set) var isActive = false
    private(set) var turnCountdown: Int?
    private(set) var characterName: String = ""

    var isMyTurn: Bool {
        currentTurnUserId == currentUserId
    }

    // MARK: - Private

    private let repository: any AvatarRepository
    private let logger = BayitLogger(category: "SharedInteractionVM")
    private var currentUserId: String = ""
    private var partyId: String = ""
    private var countdownTask: Task<Void, Never>?
    private var refreshTask: Task<Void, Never>?

    init(repository: any AvatarRepository) {
        self.repository = repository
    }

    // MARK: - Start Session

    func startSharedInteraction(
        partyId: String,
        contentId: String,
        momentTimestamp: Double,
        characterName: String,
        userId: String,
        profileId: String,
        avatarId: String,
        displayName: String
    ) async {
        self.partyId = partyId
        currentUserId = userId
        self.characterName = characterName

        do {
            let response = try await repository.startSharedInteraction(
                partyId: partyId,
                contentId: contentId,
                momentTimestamp: momentTimestamp,
                characterName: characterName,
                profileId: profileId,
                avatarId: avatarId,
                displayName: displayName
            )
            sessionId = response.id
            isActive = true
            exchanges = []
            logger.info("Shared interaction started: \(response.id)")
        } catch {
            logger.error("Failed to start shared interaction: \(error)")
        }
    }

    // MARK: - Send Message

    func sendMessage(
        _ text: String,
        addressedCharacter: String? = nil
    ) async {
        guard let sessionId, isActive, isMyTurn else { return }
        isSending = true
        defer { isSending = false }

        do {
            let response = try await repository.sendSharedMessage(
                partyId: partyId,
                sessionId: sessionId,
                message: text,
                addressedCharacter: addressedCharacter
            )

            let userExchange = DialogueExchange(
                speaker: "user",
                messageText: text,
                audioUrl: nil,
                animatedVideoUrl: nil,
                characterName: nil,
                addressedTo: addressedCharacter,
                reactionTo: nil,
                participantUserId: currentUserId,
                participantName: nil
            )
            exchanges.append(userExchange)

            for exchange in response.exchanges {
                let dialogueExchange = DialogueExchange(
                    speaker: exchange.speaker,
                    messageText: exchange.messageText,
                    audioUrl: exchange.audioUrl,
                    animatedVideoUrl: exchange.animatedVideoUrl,
                    characterName: exchange.characterName,
                    addressedTo: nil,
                    reactionTo: exchange.reactionTo,
                    participantUserId: nil,
                    participantName: nil
                )
                exchanges.append(dialogueExchange)
            }

            logger.info("Shared message sent")
        } catch {
            logger.error("Failed to send shared message: \(error)")
        }
    }

    // MARK: - Handle Events

    func handleInteractionEvent(_ event: [String: Any]) {
        guard let eventType = event["event"] as? String else { return }

        switch eventType {
        case "turn_change":
            currentTurnUserId = event["current_turn_user_id"] as? String
            turnsCompleted = event["turns_completed"] as? Int ?? turnsCompleted
            startCountdown()
        case "turn_skipped":
            currentTurnUserId = event["next_user_id"] as? String
            startCountdown()
        case "turn_warning":
            turnCountdown = event["seconds_remaining"] as? Int
        case "participant_joined":
            refreshState()
        case "interaction_end":
            isActive = false
            countdownTask?.cancel()
        default:
            break
        }
    }

    // MARK: - End Session

    func endSharedInteraction() async {
        guard let sessionId else { return }
        do {
            _ = try await repository.endSharedInteraction(
                partyId: partyId,
                sessionId: sessionId
            )
            isActive = false
            countdownTask?.cancel()
            logger.info("Shared interaction ended")
        } catch {
            logger.error("Failed to end shared interaction: \(error)")
        }
    }

    // MARK: - Private

    private func startCountdown() {
        countdownTask?.cancel()
        turnCountdown = nil
    }

    private func refreshState() {
        guard let sessionId else { return }
        refreshTask = Task {
            do {
                let state = try await repository.getSharedInteractionState(
                    partyId: partyId,
                    sessionId: sessionId
                )
                participants = state.participants
                currentTurnUserId = state.currentTurnUserId
                turnsCompleted = state.turnsCompleted
                maxTurnsPerParticipant = state.maxTurnsPerParticipant
            } catch {
                logger.error("Failed to refresh state: \(error)")
            }
        }
    }
}
